CommonJS support (#896)
denoland/docs: runtime/fundamentals/node.md

Same hints and additional quick-fixes are provided by the Deno LSP in your
editor.

## CommonJS support

> CommonJS is a module system that predetes

Reviewer: predates? Or preceded?

[ES modules](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules).
While we firmly believe that ES modules are the future of JavaScript, there are
millions of npm libraries that are written in CommonJS and Deno offers full
support for them. Deno will automatically determine if a package is using

> CommonJS and makes it work seemlessly when imported:

Reviewer: ```suggestion
CommonJS and make it work seamlessly when imported:
```

```js, title="main.js"
import react from "npm:react";
console.log(react);
```

```shell
$ deno run -E main.js
18.3.1
```

> _`npm:react` is a CommonJS package. Deno provides a seemless way to import it._

Reviewer: ```suggestion
_`npm:react` is a CommonJS package. Deno provides a seamless way to import it._
```

Reviewer: Duplication of the word "seamless" from above reads a little oddly.

That said, Deno strongly encourages use of ES modules, and offers restriced

> support for CommonJS in user code. There are two ways to write and consume

Reviewer: ```suggestion
support for CommonJS in user code.
 
There are two ways to write and consume
```

CommonJS code

### Use `.cjs` extension

If the file extension is `.cjs` Deno will treat this module as CommonJS.

```js title="main.cjs"
const express = require("express");
```

> Deno does not look for `package.json` files and `type` option to determine if

Reviewer: ```suggestion
Deno does not look for `package.json` files or `type` options to determine if
```

Author: I believe this is actually correct - `type` is only specified in `package.json`. So you can't have `type` without `package.json` hence the "and"

the file is CommonJS or ESM.

When using CommonJS, Deno expects that dependencies will be installed manually

> and `node_modules` directory will be present. It's best to set

Reviewer: ```suggestion
and a `node_modules` directory will be present. It's best to set
```

`"nodeModulesDir": "auto"` in your `deno.json` to ensure that.

```shell
$ cat deno.json
{
  "nodeModulesDir": "auto"
}

$ deno install npm:express
Add npm:express@5.0.0

$ deno run -R -E main.cjs
[Function: createApplication] {
  application: {
    init: [Function: init],
    defaultConfiguration: [Function: defaultConfiguration],
    ...
  }
}
```

> Notice that `-R` and `-E` flags were used to give access to read and env

Reviewer: ```suggestion
`-R` and `-E` flags are used to allow permissions to read files and environment variables.
```

permissions.

> **Deno permission system is still in effect when using CommonJS modules.** It is

Reviewer: ```suggestion
**Deno's permission system is still in effect when using CommonJS modules.** It is
```

necessary to provide at least `--allow-read` permission as Deno will probe the
file system for `package.json` files and `node_modules` directory to properly
resolve CommonJS modules.

### Create `require()` manually

> Alternative option is to create an instance of `require()` function manually,

Reviewer: ```suggestion
An alternative option is to create an instance of the `require()` function manually,
```

> this can be done as such:

Reviewer: ```suggestion
```
